File: client/web/src/tests/ExternalImports.test.ts
import { describe, it, expect } from "vitest";
import {
  tryParseGoogleAuthMigration,
  tryParseAegis,
  tryParse2Fas,
} from "@/features/vault/utils/external-imports";

// --- Tiny protobuf builder (mirror of the Android test helper) ------------

class ProtoBuilder {
  private parts: Uint8Array[] = [];

  varintField(field: number, value: number): this {
    this.writeTag(field, 0);
    this.writeVarint(BigInt(value));
    return this;
  }

  stringField(field: number, value: string): this {
    return this.bytesField(field, new TextEncoder().encode(value));
  }

  bytesField(field: number, value: Uint8Array): this {
    this.writeTag(field, 2);
    this.writeVarint(BigInt(value.length));
    this.parts.push(value);
    return this;
  }

  build(): Uint8Array {
    const total = this.parts.reduce((n, p) => n + p.length, 0);
    const out = new Uint8Array(total);
    let offset = 0;
    for (const p of this.parts) { out.set(p, offset); offset += p.length; }
    return out;
  }

  private writeTag(field: number, wire: number) {
    this.writeVarint(BigInt((field << 3) | wire));
  }

  private writeVarint(v: bigint) {
    const bytes: number[] = [];
    let value = v;
    while ((value & ~0x7fn) !== 0n) {
      bytes.push(Number((value & 0x7fn) | 0x80n));
      value >>= 7n;
    }
    bytes.push(Number(value));
    this.parts.push(new Uint8Array(bytes));
  }
}

function u8ToBase64(bytes: Uint8Array): string {
  let binary = "";
  for (const b of bytes) binary += String.fromCharCode(b);
  return btoa(binary);
}

// Base32 decoder (for the test to produce expected `secret` values).
const BASE32 = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";
function base32Decode(s: string): Uint8Array {
  const cleaned = s.replace(/=/g, "").toUpperCase();
  let bits = "";
  for (const c of cleaned) {
    const v = BASE32.indexOf(c);
    if (v < 0) throw new Error(`bad base32: ${c}`);
    bits += v.toString(2).padStart(5, "0");
  }
  const bytes: number[] = [];
  for (let i = 0; i + 8 <= bits.length; i += 8) {
    bytes.push(parseInt(bits.slice(i, i + 8), 2));
  }
  return new Uint8Array(bytes);
}

// --- Google Authenticator -------------------------------------------------

describe("tryParseGoogleAuthMigration", () => {
  it("decodes two entries with different algorithms and digit counts", () => {
    const secret1 = base32Decode("JBSWY3DPEHPK3PXP");
    const secret2 = base32Decode("NBSWY3DPFQQHO33SNRSCC");

    const otp1 = new ProtoBuilder()
      .bytesField(1, secret1)
      .stringField(2, "alice@example.com")
      .stringField(3, "Example")
      .varintField(4, 1) // SHA1
      .varintField(5, 1) // SIX
      .varintField(6, 2) // TOTP
      .build();

    const otp2 = new ProtoBuilder()
      .bytesField(1, secret2)
      .stringField(2, "bob")
      .stringField(3, "")
      .varintField(4, 2) // SHA256
      .varintField(5, 2) // EIGHT
      .varintField(6, 2) // TOTP
      .build();

    const payload = new ProtoBuilder()
      .bytesField(1, otp1)
      .bytesField(1, otp2)
      .varintField(2, 1)
      .build();

    const uri = `otpauth-migration://offline?data=${encodeURIComponent(u8ToBase64(payload))}`;
    const entries = tryParseGoogleAuthMigration(uri);

    expect(entries).not.toBeNull();
    expect(entries).toHaveLength(2);
    expect(entries![0]).toMatchObject({
      issuer: "Example",
      accountName: "alice@example.com",
      secret: "JBSWY3DPEHPK3PXP",
      algorithm: "SHA1",
      digits: 6,
    });
    expect(entries![1]).toMatchObject({
      accountName: "bob",
      secret: "NBSWY3DPFQQHO33SNRSCC",
      algorithm: "SHA256",
      digits: 8,
    });
  });

  it("splits Issuer:account when the issuer field is empty", () => {
    const otp = new ProtoBuilder()
      .bytesField(1, base32Decode("JBSWY3DPEHPK3PXP"))
      .stringField(2, "GitHub:octocat")
      .stringField(3, "")
      .varintField(4, 1).varintField(5, 1).varintField(6, 2)
      .build();
    const payload = new ProtoBuilder().bytesField(1, otp).build();
    const uri = `otpauth-migration://offline?data=${encodeURIComponent(u8ToBase64(payload))}`;

    const entries = tryParseGoogleAuthMigration(uri);
    expect(entries).toHaveLength(1);
    expect(entries![0]).toMatchObject({ issuer: "GitHub", accountName: "octocat" });
  });

  it("strips redundant issuer prefix from name when it matches the issuer field", () => {
    const otp = new ProtoBuilder()
      .bytesField(1, base32Decode("JBSWY3DPEHPK3PXP"))
      .stringField(2, "GitHub:octocat") // redundant prefix
      .stringField(3, "GitHub")          // explicit issuer matches
      .varintField(4, 1).varintField(5, 1).varintField(6, 2)
      .build();
    const payload = new ProtoBuilder().bytesField(1, otp).build();
    const uri = `otpauth-migration://offline?data=${encodeURIComponent(u8ToBase64(payload))}`;

    const entries = tryParseGoogleAuthMigration(uri)!;
    expect(entries[0]).toMatchObject({ issuer: "GitHub", accountName: "octocat" });
  });

  it("keeps colons in account when the prefix doesn't match the issuer", () => {
    // Legit colon inside the account — don't misread it as "Issuer:account".
    const otp = new ProtoBuilder()
      .bytesField(1, base32Decode("JBSWY3DPEHPK3PXP"))
      .stringField(2, "work:alice")
      .stringField(3, "Company")
      .varintField(4, 1).varintField(5, 1).varintField(6, 2)
      .build();
    const payload = new ProtoBuilder().bytesField(1, otp).build();
    const uri = `otpauth-migration://offline?data=${encodeURIComponent(u8ToBase64(payload))}`;

    const entries = tryParseGoogleAuthMigration(uri)!;
    expect(entries[0]).toMatchObject({ issuer: "Company", accountName: "work:alice" });
  });

  it("filters out HOTP entries", () => {
    const otp = new ProtoBuilder()
      .bytesField(1, base32Decode("JBSWY3DPEHPK3PXP"))
      .stringField(2, "counter-based")
      .stringField(3, "Bank")
      .varintField(4, 1).varintField(5, 1)
      .varintField(6, 1) // HOTP
      .build();
    const payload = new ProtoBuilder().bytesField(1, otp).build();
    const uri = `otpauth-migration://offline?data=${encodeURIComponent(u8ToBase64(payload))}`;

    expect(tryParseGoogleAuthMigration(uri)).toEqual([]);
  });

  it("returns null for non-migration URIs", () => {
    expect(tryParseGoogleAuthMigration("otpauth://totp/foo?secret=JBSW")).toBeNull();
    expect(tryParseGoogleAuthMigration("https://example.com")).toBeNull();
  });
});

// --- Aegis ----------------------------------------------------------------

describe("tryParseAegis", () => {
  it("parses a plain export", () => {
    const json = {
      version: 1,
      header: { slots: null, params: null },
      db: {
        version: 3,
        entries: [
          {
            type: "totp",
            name: "alice@example.com",
            issuer: "Example",
            info: {
              secret: "jbswy3dpehpk3pxp",
              algo: "sha1",
              digits: 6,
              period: 30,
            },
          },
          {
            type: "hotp",
            name: "counter",
            issuer: "Bank",
            info: { secret: "X", algo: "SHA1", digits: 6, counter: 0 },
          },
        ],
      },
    };
    const entries = tryParseAegis(json);
    expect(entries).not.toBeNull();
    expect(entries).toHaveLength(1); // HOTP filtered
    expect(entries![0]).toMatchObject({
      issuer: "Example",
      accountName: "alice@example.com",
      secret: "JBSWY3DPEHPK3PXP",
      algorithm: "SHA1",
      digits: 6,
      period: 30,
    });
  });

  it("rejects shapes without header", () => {
    const json = { db: { entries: [{ type: "totp", info: { secret: "X" } }] } };
    expect(tryParseAegis(json)).toBeNull();
  });
});

// --- 2FAS -----------------------------------------------------------------

describe("tryParse2Fas", () => {
  it("parses an export with otp subobject", () => {
    const json = {
      schemaVersion: 4,
      appVersionCode: 500000,
      services: [
        {
          secret: "JBSWY3DPEHPK3PXP",
          name: "Example",
          otp: {
            account: "alice@example.com",
            issuer: "Example",
            digits: 6,
            period: 30,
            algorithm: "SHA1",
          },
        },
      ],
    };
    const entries = tryParse2Fas(json);
    expect(entries).toHaveLength(1);
    expect(entries![0]).toMatchObject({
      issuer: "Example",
      accountName: "alice@example.com",
      secret: "JBSWY3DPEHPK3PXP",
    });
  });

  it("falls back to name when otp.account is missing", () => {
    const json = {
      schemaVersion: 3,
      services: [{ secret: "JBSWY3DPEHPK3PXP", name: "Legacy Service" }],
    };
    const entries = tryParse2Fas(json);
    expect(entries).toHaveLength(1);
    expect(entries![0]).toMatchObject({
      issuer: "Legacy Service",
      accountName: "Legacy Service",
    });
  });

  it("rejects shapes without schemaVersion", () => {
    const json = { services: [{ secret: "X" }] };
    expect(tryParse2Fas(json)).toBeNull();
  });
});

File: client/android/core/src/test/java/io/github/pstanar/pstotp/core/model/ExternalImportsTest.kt
package io.github.pstanar.pstotp.core.model

import io.github.pstanar.pstotp.core.crypto.Base32
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import java.util.Base64

class ExternalImportsTest {

    // --- Google Authenticator ---------------------------------------------

    @Test
    fun `Google Auth migration payload decodes both entries`() {
        // Hand-built MigrationPayload:
        //   OtpParameters { secret=<raw>, name="alice@example.com", issuer="Example",
        //                   algorithm=SHA1, digits=SIX, type=TOTP }
        //   OtpParameters { secret=<raw>, name="bob", issuer="", algorithm=SHA256,
        //                   digits=EIGHT, type=TOTP }
        val secret1 = Base32.decode("JBSWY3DPEHPK3PXP")
        val secret2 = Base32.decode("NBSWY3DPFQQHO33SNRSCC")

        val otp1 = protoBuilder()
            .bytesField(1, secret1)
            .stringField(2, "alice@example.com")
            .stringField(3, "Example")
            .varintField(4, 1) // SHA1
            .varintField(5, 1) // SIX
            .varintField(6, 2) // TOTP
            .build()

        val otp2 = protoBuilder()
            .bytesField(1, secret2)
            .stringField(2, "bob")
            .stringField(3, "")
            .varintField(4, 2) // SHA256
            .varintField(5, 2) // EIGHT
            .varintField(6, 2) // TOTP
            .build()

        val payload = protoBuilder()
            .bytesField(1, otp1)
            .bytesField(1, otp2)
            .varintField(2, 1) // version
            .build()

        val uri = "otpauth-migration://offline?data=" +
            Base64.getEncoder().encodeToString(payload)

        val entries = ExternalImports.tryParseGoogleAuthMigration(uri)
        assertNotNull(entries)
        assertEquals(2, entries!!.size)

        assertEquals("Example", entries[0].issuer)
        assertEquals("alice@example.com", entries[0].accountName)
        assertEquals(Base32.encode(secret1), entries[0].secret)
        assertEquals("SHA1", entries[0].algorithm)
        assertEquals(6, entries[0].digits)

        assertEquals("bob", entries[1].accountName)
        assertEquals(Base32.encode(secret2), entries[1].secret)
        assertEquals("SHA256", entries[1].algorithm)
        assertEquals(8, entries[1].digits)
    }

    @Test
    fun `Google Auth migration splits Issuer_Account in name field`() {
        // Some exports put "Issuer:account" in name with empty issuer field.
        val secret = Base32.decode("JBSWY3DPEHPK3PXP")
        val otp = protoBuilder()
            .bytesField(1, secret)
            .stringField(2, "GitHub:octocat")
            .stringField(3, "")
            .varintField(4, 1)
            .varintField(5, 1)
            .varintField(6, 2)
            .build()
        val payload = protoBuilder().bytesField(1, otp).build()
        val uri = "otpauth-migration://offline?data=" +
            Base64.getEncoder().encodeToString(payload)

        val entries = ExternalImports.tryParseGoogleAuthMigration(uri)!!
        assertEquals(1, entries.size)
        assertEquals("GitHub", entries[0].issuer)
        assertEquals("octocat", entries[0].accountName)
    }

    @Test
    fun `Google Auth strips redundant issuer prefix from name when it matches`() {
        val otp = protoBuilder()
            .bytesField(1, Base32.decode("JBSWY3DPEHPK3PXP"))
            .stringField(2, "GitHub:octocat")  // redundant prefix
            .stringField(3, "GitHub")           // explicit issuer matches
            .varintField(4, 1).varintField(5, 1).varintField(6, 2)
            .build()
        val uri = buildMigrationUri(otp)

        val entries = ExternalImports.tryParseGoogleAuthMigration(uri)!!
        assertEquals("GitHub", entries[0].issuer)
        assertEquals("octocat", entries[0].accountName)
    }

    @Test
    fun `Google Auth keeps colons in account when prefix does not match issuer`() {
        // Legit colon inside the account — don't misread it as "Issuer:account".
        val otp = protoBuilder()
            .bytesField(1, Base32.decode("JBSWY3DPEHPK3PXP"))
            .stringField(2, "work:alice")
            .stringField(3, "Company")
            .varintField(4, 1).varintField(5, 1).varintField(6, 2)
            .build()
        val uri = buildMigrationUri(otp)

        val entries = ExternalImports.tryParseGoogleAuthMigration(uri)!!
        assertEquals("Company", entries[0].issuer)
        assertEquals("work:alice", entries[0].accountName)
    }

    @Test
    fun `Google Auth migration skips HOTP entries`() {
        val otp = protoBuilder()
            .bytesField(1, Base32.decode("JBSWY3DPEHPK3PXP"))
            .stringField(2, "counter-based")
            .stringField(3, "Bank")
            .varintField(4, 1)
            .varintField(5, 1)
            .varintField(6, 1) // HOTP — should be filtered out
            .build()
        val payload = protoBuilder().bytesField(1, otp).build()
        val uri = "otpauth-migration://offline?data=" +
            Base64.getEncoder().encodeToString(payload)

        val entries = ExternalImports.tryParseGoogleAuthMigration(uri)!!
        assertTrue(entries.isEmpty())
    }

    @Test
    fun `returns null for non-migration URIs`() {
        assertNull(ExternalImports.tryParseGoogleAuthMigration("otpauth://totp/foo?secret=JBSW"))
        assertNull(ExternalImports.tryParseGoogleAuthMigration("https://example.com"))
    }

    // --- Aegis -----------------------------------------------------------

    @Test
    fun `Aegis plain export parses`() {
        val json = JSONObject(
            """
            {
              "version": 1,
              "header": { "slots": null, "params": null },
              "db": {
                "version": 3,
                "entries": [
                  {
                    "type": "totp",
                    "name": "alice@example.com",
                    "issuer": "Example",
                    "info": {
                      "secret": "jbswy3dpehpk3pxp",
                      "algo": "sha1",
                      "digits": 6,
                      "period": 30
                    }
                  },
                  {
                    "type": "hotp",
                    "name": "counter",
                    "issuer": "Bank",
                    "info": { "secret": "NBSWY3DPFQQHO33SNRSCC", "algo": "SHA1", "digits": 6, "counter": 0 }
                  }
                ]
              }
            }
            """.trimIndent(),
        )
        val entries = ExternalImports.tryParseAegis(json)
        assertNotNull(entries)
        assertEquals(1, entries!!.size) // HOTP filtered
        assertEquals("Example", entries[0].issuer)
        assertEquals("alice@example.com", entries[0].accountName)
        assertEquals("JBSWY3DPEHPK3PXP", entries[0].secret) // normalised to upper
        assertEquals("SHA1", entries[0].algorithm)
    }

    @Test
    fun `Aegis rejects shape without header`() {
        val json = JSONObject(
            """{ "db": { "entries": [ { "type": "totp", "info": { "secret": "X" } } ] } }""",
        )
        assertNull(ExternalImports.tryParseAegis(json))
    }

    // --- 2FAS ------------------------------------------------------------

    @Test
    fun `2FAS export parses`() {
        val json = JSONObject(
            """
            {
              "schemaVersion": 4,
              "appVersionCode": 500000,
              "services": [
                {
                  "secret": "JBSWY3DPEHPK3PXP",
                  "name": "Example",
                  "otp": {
                    "account": "alice@example.com",
                    "issuer": "Example",
                    "digits": 6,
                    "period": 30,
                    "algorithm": "SHA1"
                  }
                }
              ]
            }
            """.trimIndent(),
        )
        val entries = ExternalImports.tryParse2Fas(json)
        assertNotNull(entries)
        assertEquals(1, entries!!.size)
        assertEquals("Example", entries[0].issuer)
        assertEquals("alice@example.com", entries[0].accountName)
        assertEquals("JBSWY3DPEHPK3PXP", entries[0].secret)
    }

    @Test
    fun `2FAS falls back to name when otp account is missing`() {
        val json = JSONObject(
            """
            {
              "schemaVersion": 3,
              "services": [ { "secret": "JBSWY3DPEHPK3PXP", "name": "Legacy Service" } ]
            }
            """.trimIndent(),
        )
        val entries = ExternalImports.tryParse2Fas(json)!!
        assertEquals(1, entries.size)
        assertEquals("Legacy Service", entries[0].issuer)
        assertEquals("Legacy Service", entries[0].accountName)
    }

    @Test
    fun `2FAS rejects shape without schemaVersion`() {
        val json = JSONObject("""{ "services": [ { "secret": "X" } ] }""")
        assertNull(ExternalImports.tryParse2Fas(json))
    }

    // --- Tiny protobuf builder used by the test ---------------------------

    private class ProtoBuilder {
        private val out = java.io.ByteArrayOutputStream()

        fun varintField(field: Int, value: Long): ProtoBuilder {
            writeTag(field, 0)
            writeVarint(value)
            return this
        }

        fun varintField(field: Int, value: Int) = varintField(field, value.toLong())

        fun stringField(field: Int, value: String): ProtoBuilder =
            bytesField(field, value.toByteArray(Charsets.UTF_8))

        fun bytesField(field: Int, value: ByteArray): ProtoBuilder {
            writeTag(field, 2)
            writeVarint(value.size.toLong())
            out.write(value)
            return this
        }

        fun build(): ByteArray = out.toByteArray()

        private fun writeTag(field: Int, wire: Int) = writeVarint(((field shl 3) or wire).toLong())

        private fun writeVarint(v: Long) {
            var value = v
            while (value and 0x7fL.inv() != 0L) {
                out.write(((value and 0x7fL) or 0x80L).toInt())
                value = value ushr 7
            }
            out.write(value.toInt())
        }
    }

    private fun protoBuilder() = ProtoBuilder()

    private fun buildMigrationUri(vararg otpParamsBytes: ByteArray): String {
        val payload = ProtoBuilder().apply {
            for (p in otpParamsBytes) bytesField(1, p)
        }.build()
        return "otpauth-migration://offline?data=" +
            Base64.getEncoder().encodeToString(payload)
    }
}
